core-backend/wasmi/src/funcs/mod.rs

```diff
@@ -542,114 +474,125 @@ where
                         delay,
                     )
                 },
-                message_id_ptr
+                |res, mut mem_ref| {
+                    let err_mid = res
+                        .map(|message_id| LengthWithHash {
+                            hash: message_id.into(),
+                            ..Default::default()
+                        })
+                        .unwrap_or_else(|length| LengthWithHash {
+                            length,
+                            ..Default::default()
+                        });
+
+                    mem_ref.write_memory_as(err_mid_ptr, err_mid)
+                },
```

Comment:
The code looks repetitive too. Could we have a second method, for instance `call_fallible_hash`, that executes this logic inside its body?